### ISSUE TITLE
Roadmap? And some additional questions / comments

### ISSUE THREAD

**Issue by survirtual** [makepad/makepad#125]
First, I want to say this project is brilliant.  Well done.  Going through the codebase, I was thrilled to see the techniques used for geometry gen, shader compilation for multiple targets, shader to rust live data binding, contained text rendering / glyph generation / etc.  This has captured many objectives I was targeting myself, so using this work as a base shaves years off my dev.

A few questions / comments:

- this could become THE next gen framework for web & standalone apps.  I see there is a lot of work being done on the live studio editor, but is the intention to make everything (renderer & components) modular and standalone?  I am less interested in a fully functional editor with live code changes (vscode works fine for me atm), and more interested in having the simple building blocks to create an app

- it seems like there is a lot of work that needs to be done regarding usability and documentation.  The renderer is brilliant.  The vision for the live code editor is as well -- it's magical seeing code changes reflected so responsively.  The core rendering tech & using shaders defined on a per-component basis is really what I'm after however.

- is there a publicly available roadmap?  I'd like to make contributions around usability and providing a react / vue / angular-like component experience, and perhaps implementing more core components / a component library needed to make an app.  My dream is being able to nearly eliminate boilerplate for UI and having absolute focus on the domain being implemented.  I'd like to add more macros for generating components & declaring component dependencies with less boilerplate.

- is there a reason in particular there aren't really any deps / a lot of code & techniques were directly inserted?  Example is window management for multiple platforms, instead of using winit it looks like the code was copied in?  It sort of breaks the modularity / flexibility of the system when the real meat is in the rendering, so I was curious about the reasoning.  Seems like not using cargo really will make common efforts hard / more complex for this project.  Any plans to be more integrated with the existing cargo infrastructure, ex configurable window management, cg math choices, etc.

- plans for accessibility?

Will have more thoughts, is writing them here appropriate?

**Comment by makepaddev** [comment]
Thanks! It's been a long road (nearly a decade now) but it's finally coming together.
Everythign is architected componentised and MIT licensed. We have (in the live branch)
- the render crate, which has everything to open a window / draw text and shader and platform abstraction
- the component crate containing the base UI kit (buttons, textinput) WIP
- makepad studio component crate containing the docksystem/tabs/etc
- makepad studio itself containing the code editor and collab
- makepad studio design (submodule atm) which is the only commercially licensed bit containing the visual design tooling

So to answer your question, its all very much modularised to fit your usecases.
Roadmap wise, we should probably make one. For now our roadmap is:
- make a first version of the FULL vision (with the visual designsuite) (3mo)
- create a documented OSS release (push to crates.io + docs) of the the whlole system (next 9mo)

In Rust the boilerplate we have, really has a function. I've tried very hard to reduce it but i think we are going to simply autogenerate it from the IDE to improve UX. The procmacros we use in the live branch (derive(LIve) already generates a lot of boilerplate, to the max i could do.

We don't have deps 'copied in' like winit, its really a from scratch implementation. The reason its integrated into the 'render' crate is that its simply fairly tightly integrated and separating this out more serves little purpose. It's already very contained. For instance, how would you see separating 'windowing' from 'rendering' on web.. This is why winit for us failed as well, its the wrong abstraction level.

**Comment by makepaddev** [comment]
Similarly, most of the Rust ecosystem crates (serde for instance) really throw compiletime/dependency count to the wind. And since we are building a compile-on-user-machine designsuite, as well as a foundation for people building visual applications with Rust, it seemed like a perfect moment to de-cruft the ecosystem a bit.
And by this i mean we pretty much write everything ourselves. Sofar it's kept our compiletime really tight, and i don't see this approach changing. It's both fun for us to do, as well as very functional.

**Comment by makepaddev** [comment]
So to answer your question: No we will not decouple our rendering from window management. Our goal is to provide a unified layer on all platforms including web. Nor will we separate out our shadercompiler from the rest. It serves no purpose, in fact it makes it harder for us to build our product. Accessibility we will provide on each platform specific to the platform. So on web we'd have to spawn ghost-DOM trees, and on macos you have to make nsview trees. On windows / linux i have no idea yet.

**Comment by makepaddev** [comment]
I strongly believe that overcomponentisation is a big cause of the dependency-disease that infects most of Rust space, but also JS with NPM. Developers have forgotten to write even the most basic of functions (like left-pad). If you keep a (self written) dependency inside of its own usecase-domain, you can actually write code that you need for a particular problem and stop a component from bloating with everyone elses usecases dragging everyones compiletime, downloadtime and depencency count to hell. And this means you get to compile our  complete desktop IDE stack in 10-15 seconds with all dependencies, releasemode on one of those new macbooks.

**Comment by survirtual** [comment]
Okay, thanks for the fast responses!  I understand your thinking.  I appreciate all the work being done and sharing it publicly.  Apologies for assuming the window management code was ripped from winit, I saw some comments referring to it in the codebase but didn't do a diff compare.

So for web windowing etc, winit has this example: https://github.com/rust-windowing/winit/blob/master/examples/web.rs.  I do think it could be beneficial to use a more unified rust interface so that people outside this ecosystem have the ability to understand the contract.  The bevy engine does this pretty well with plugins, https://bevyengine.org/learn/book/getting-started/plugins/

Making rendering, window management, etc composable doesn't compromise your objective for a unified layer, by offering sane defaults most users needs will be met.  But in the event I want to directly tap into window events or otherwise manage windows myself, it's easy to do with a common contract.

It also helps external devs understanding your codebase.  It divides the concerns of the application which makes it easier to digest the code, allowing more contributions and bug fixes.

**Comment by survirtual** [comment]
> I strongly believe that overcomponentisation is a big cause of the dependency-disease that infects most of Rust space, but also JS with NPM. Developers have forgotten to write even the most basic of functions (like left-pad). If you keep a (self written) dependency inside of its own usecase-domain, you can actually write code that you need for a particular problem and stop a component from bloating with everyone elses usecases dragging everyones compiletime, downloadtime and depencency count to hell. And this means you get to compile our complete desktop IDE stack in 10-15 seconds with all dependencies, releasemode on one of those new macbooks.

We are certainly on the same page here.  There is a right way to do dependencies and a wrong way.  I think rust has learned a lot from the failings of npm and the js ecosystem, and there is a lot of value in doing deps properly.  makepad certainly does compile blazing fast on my m1, which is very nice.

I am approaching this from outside your codebase, so please take no offense!  Just offering some of my perspective as an outsider that understands what you have here / where it can go.

**Comment by makepaddev** [comment]
About winit, i definitely used it to learn the first steps on macos or how to call the platform API's on windows. However for the rest its pretty much a complete rewrite and has grown to have a lot more features afaik. But i don't keep track there.  And yea you do have a point, and i've thought about this as well. How to modularise this more. I've started by atleast splitting things out into submodules more and writing some 'trait' contracts there. However every time i try 'web' just breaks all of it, or it reduces to some meaningless shill.
As for how my web code talks to the JS code i use a message passing system. Not the direct-call stuff you see in wasm-bindgen or that winit example. It works extremely well to contain complexity, but also kinda marries the rendering system to the windowing system. Similarly in webXR these things are very closely related. 

However maybe i'll see a path where this separates out. However i wouldn't want that for the purpose of broadening the use of our tech outside of our usecase tho, i need that compilespeed more.

**Comment by makepaddev** [comment]
I'm not offended btw, just a spirited individual. And i do see your points.

**Comment by makepaddev** [comment]
Makepad is fully componentised, however not been published to crates io as to not get lost in any opensource uptake of its individual components. Every layer and system we have is a potential complete space on its own with endless github discussions as to its design or features. We wouldn't even get out the door if we tried to create open source ecosystems for each part. We have to approach this as a complete system, where the individual components serve that exact purpose.

**Comment by makepaddev** [comment]
We aim to create a new 'visual basic/delphi' for Rust, with deploy to web. And much fancier design features with shader flowgraphs and so on. I don't have the time and energy to manage a community around our replacement for serde, or winit or any of the other components independent of that goal.

**Comment by makepaddev** [comment]
However i do hope that with this strategy we can create something thats full featured enough to be valuable. I'm quite hopeful of that. But we'll know soon enough.

**Comment by survirtual** [comment]
Just to be clear, this codebase is objectively valuable as is.  I was in the process of writing a gui engine directly with webgpu because the existing ecosystem is inadequate with consideration to wasm & true, performant cross-platform.  Using webgpu also has disadvantages (in newness and support) that would exclude a large population, which I was okay with for my purposes.

Looking at existing libs, there are so many compromises needing to be made.   So a tunnel forms where you have to do everything yourself it seems like, without being locked in to any specific platform.  The time cost of that is years, and that is a lot to swallow on top of the things I'm implementing.

UI should be extremely beautiful, lightning fast, and totally mobile with the tech stacks possible now.  People implementing apps on specific domains should be largely insulated from as many barriers to a reaching their ideal use cases as possible, while easily being platform agnostic.  But there is a lot of laziness and greed.  This codebase is not lazy, it makes no assumptions that other systems are done correctly and just does it well itself.  This has a lot of advantages and while maybe not be ideal in a more functional larger ecosystem, may be necessary right now given the general cross platform code smells that exist.  

It is a large undertaking though, and I think many oss people would help lighten the load with a more modularized codebase in the future.  What's being done here seems more comparable to a web browser (that is truly XR minded) or an OS display layer rather than VB or Delphi, but all these things in any case traditionally have been walled gardens or esoteric blackbox monoliths.  That said, building a monolith first to get the full vision implemented & rapidly available for a new pattern is probably the only real way with more limited resources.

Long story short, I will spend more time playing and building with what's available.  I have a better understanding where things are at & going now.  I hope it will be on the agenda once more stabilized to organize the project side of things to allow an easier flow of external contributors, such as:

- a project board showing what's being worked on, feature targets for releases, etc. so new people don't duplicate efforts / can directly contribute where needed to accelerate dev
- architecture docs connecting the high level dots
- generally available roadmap

This is a big undertaking.  Thanks again & excited to see it evolve.

**Comment by makepaddev** [comment]
We'll get that roadmap up soon. Architecture docs are planned for mid this year. We hope to deliver on most of what you'd need to build your application this year. We'll add webGPU support once safari has it rolled out. Last thing i like is being 'cutting edge' with webAPIs, voids the whole idea of web.

**Comment by makepaddev** [comment]
In terms of it being more like a webbrowser than say, visual basic. Yea it's true its a bit more fundamental in that sense. However it's also much more compact than a webbrowser. Because otherwise we'd take 45 minutes to compile :)

**Comment by makepaddev** [comment]
And yea our goal is to complete the vision first before taking on 'drag' in that sense. I can't have arguments about redoing our windowing api independently of our goals. Not until after we reach that 'completeness' for the stack we are building.

**Comment by survirtual** [comment]
Sounds great.

On the topic of the proc_macros, any reason [quote](https://docs.rs/quote/latest/quote/#) wasn't used to generate token streams?  Doesn't add much in way of compile time.  Makes it much nicer to write complex proc macros.  Gives full syntax highlighting for generated code and eliminates the need for escaping strings, among other conveniences.

Here's an example from some code I put together to generate an arbitrary base from an alphabet (notice the quote! and later generated of a TokenStream):

```rust
fn impl_base_x(alphabet: &[u8]) -> TokenStream {
    let base = alphabet.len();
    let alphabet_tokens = alphabet.iter();

    if base > 255 {
        panic!("Alphabet is too long");
    }


    let mut base_map: [usize; 256] = [255; 256];
    let mut i = 0;
    while i < base {
        let x = alphabet[i] as char;
        let xc = alphabet[i] as usize;

        if base_map[xc] != 255 {
            panic!(format!("{} is ambiguous", x));
        }

        base_map[xc] = i;
        i += 1;
    }

    let factor = (base as f32).ln() / (256.0_f32).ln();
    let i_factor = (256.0_f32).ln() / (base as f32).ln();

    let gen = quote! {
        const ALPHABET:[u8; #base] = [#(#alphabet_tokens),*];
        const BASE: usize = #base;
        const BASE_MAP: [usize; 256] = [#(#base_map),*];
        const LEADER: char = ALPHABET[0] as char;
        const FACTOR: f32 = #factor;
        const I_FACTOR: f32 = #i_factor;

        pub fn encode(source: &[u8]) -> Result<String, String> {
            if source.len() == 0 {
                return Ok("".to_owned());
            }
    
            let mut zeroes = 0;
            let mut length = 0;
            let mut p_begin = 0;
            let p_end = source.len();
    
            while p_begin != p_end && source[p_begin] == 0 {
                p_begin += 1;
                zeroes += 1;
            }
    
            let size = (((p_end - p_begin) as f32 * I_FACTOR + 1.0).trunc() as usize) >> 0;
    
            let mut b_58 = vec![0; size];
    
            while p_begin != p_end {
                let mut carry = source[p_begin] as u32;
    
                let mut i = 0;
                let mut ij = size as i32 - 1;
                while (carry != 0 || i < length) && (ij != -1) {
                    carry += (256 * b_58[ij as usize] as u32) >> 0;
                    b_58[ij as usize] = ((carry % BASE as u32) >> 0) as u8;
                    carry = (carry / BASE as u32) >> 0;
    
                    ij -= 1;
                    i += 1;
                }
    
                if carry != 0 {
                    return Err("Non-zero carry".to_owned());
                }
                length = i;
                p_begin += 1;
            }
    
            // Skip leading zeroes in base58 result.
            let mut it = size - length;
            while it != size && b_58[it] == 0 {
                it += 1;
            }
    
            // Translate the result into a string.
            let mut string = LEADER.to_string().repeat(zeroes);
            while it < size {
                string += &(ALPHABET[b_58[it] as usize] as char).to_string();
              it += 1;
            }
    
            Ok(string)
        }

        pub fn decode(source: &str) -> Result<Vec<u8>, String> {
            if source.len() == 0 {
                return Ok(vec!());
            }
    
            let bytes = source.as_bytes();
    
            let mut psz = 0;
    
            if bytes[psz] == b' ' {
                return Err(format!("Non-base {} character", BASE));
            }
    
            // Skip and count leading '1's.
            let mut zeroes = 0;
            let mut length = 0;
            while psz < bytes.len() && bytes[psz] == LEADER as u8 {
                zeroes += 1;
                psz += 1;
            }
    
            // Allocate enough space in big-endian base256 representation.
            let size = (((bytes.len() - psz) as f32 * FACTOR).trunc() as usize + 1) >> 0; // log(58) / log(256), rounded up.
            let mut b256: Vec<u8> = vec![0; size];
    
            while psz < bytes.len() && bytes[psz] != b' ' {
                let mut carry = BASE_MAP[bytes[psz] as usize] as u32;
    
                if carry == 255 {
                    return Err("Invalid character".to_owned());
                }
    
                let mut i = 0;
                let mut ij = (size - 1) as i32;
                while (carry != 0 || i < length) && (ij != -1) {
                    carry += (BASE as u32 * (b256[ij as usize] as u32)) >> 0;
                    b256[ij as usize] = ((carry % 256) >> 0) as u8;
                    carry = (carry / 256) >> 0;
    
                    ij -= 1;
                    i += 1;
                }
                if carry != 0 {
                    return Err("Non-zero carry".to_owned());
                }
                length = i;
                psz += 1;
            }

            // Skip trailing spaces.
            if psz < bytes.len() && bytes[psz] == b' ' {
                return Err(format!("Non-base {} character", BASE));
            }
            
            // Skip leading zeroes in b256.
            let mut it = size - length;
            while it != size && b256[it] == 0 {
                it += 1;
            }
    
            let mut vch: Vec<u8> = vec![0; zeroes + (size - it)];
    
            let mut j = zeroes;
            while it != size {
                vch[j] = b256[it];
                j += 1;
                it += 1;
            }
    
            return Ok(vch);
        }
    };
    TokenStream::from(gen)
}
```

**Comment by makepaddev** [comment]
So, the reason i like using my 'TokenBuilder' over quote is that i find it much easier to read what code is generated (string color) and whats the outer code. Now if you had a very smart syntaxhighlighter there  to mark the quote!{} blocks differently i might change my mind. 
Now yes, things like paren matching doesn't work and no syntax highlighting. However this turns out to be a relatively minor problem for me personally. I've also settled on an indent pattern inside the strings that makes it easier to follow.

**Comment by makepaddev** [comment]
Btw i used to write my procmacros with quote in the beginning. I found it hard to read. Switching to my 'tokenbuilder' object felt so much easier. However when you misplace a paren it has a bit of a UX cliff thats not so nice.

**Comment by makepaddev** [comment]
However i have much less issue with quote than the use of syn, which i removed at the same time. It's arguable that i could or should have kept the quote!{} mechanism. Rewriting that some day is probably a 2 day job so its not a big thing.

**Comment by survirtual** [comment]
Okay, as long as you are aware of the crate.

I'm messing with live component registration from external crates, and debugging / reading some of these proc_macros as strings isn't great.

**Comment by makepaddev** [comment]
Yea i agree with that. Putting in println!("hi") in the generated code is not very nice.

**Comment by makepaddev** [comment]
If i would use a mechanism 'like' quote i'd make my own. Quote adds 3 deps of its own again for no reason. It's not hard to write.

**Comment by survirtual** [comment]
I think for you there is not much that is hard to write.

Using quote or roll-your-own quote would make writing, reading, and maintaining the proc_macros a lot easier for people less skillful.

As an example, needed to see which child node was a problem, so made a mod to derive_live_registry.rs:
```rust
            tb.add(" impl LiveApply for ").ident(&struct_name).add(" {");
            tb.add("     fn apply(&mut self, cx: &mut Cx, apply_from: ApplyFrom, index: usize, nodes: &[LiveNode]) -> usize {");
            tb.add("         if let Some(file_id) = apply_from.file_id() {");
            tb.add("             let mut registry = cx.registries.get_or_create::<").ident(&registry).add(">();");
            tb.add("             let generation = cx.live_registry.borrow().file_id_to_file(file_id).generation;");
            tb.add("             for item in registry.items.values_mut() {");
            // Adding the format to identify the problem child took me longer than I want to admit
            tb.add("                 let index = nodes.child_by_name(index, item.id).expect(&format!(\"Registry item found, but child node '{}' not. Make sure the registry is live_register called last otherwise the DSL nodes aren't populated\", item.id)[..]);");
            tb.add("                 item.live_ptr = Some(LivePtr {file_id, index: index as u32, generation})");
            tb.add("             }");
            tb.add("         }");
            tb.add("         nodes.skip_node(index)");
            tb.add("     }");
            tb.add(" }");
```

**Comment by makepaddev** [comment]
Yea fair point, i was having issues there myself as well. I'll contemplate it, might replace it with my own quote in the near future.

**Comment by survirtual** [comment]
Okay, great.  I am happy to see you are open to discussion and consideration of these things.

You are building an amazing thing that I think will have a large population in the future.  What I am saying is not for you for for your success; keep building and success is guaranteed for you.  It is for the people who have to maintain your codebase after you are done with it because it is successful and valuable.

In general with my development, it has become trivial to build anything I imagine.  But then I consider the entire lifecycle of the application, until other people are maintaining a codebase, and I consider from the perspective of new developers trying to understand / learn from / maintain / fix / upgrade.  I have found it is more efficient for the lifecycle to make code less than optimal if it improves readability & understanding.  So I try to write code now in compassion to the people that follow the wake of success.

I don't mean to lecture, I am very grateful for what you're building & am just interested in its longevity.

**Comment by makepaddev** [comment]
Yea i was just looking at adding a quote!{} thing, it's not hard. I'll do that in a bored moment :)

**Comment by melMass** [comment]
I cam to ask mostly the same questions, very interesting answers here!!
@makepaddev So if I understand right, the work in `dev` and `live` is starting to separate the libs in disctinct crates?
I'm looking into using the UI crate for a custom app, but I'm a bit lost, I'll try to decypher infos from issues and code but if there is any guide or writeup somewhere I'd love to look at it.

**Comment by makepaddev** [comment]
Hi! live is my work-branch, and dev (will) be the branch i know compiles. We're aiming to get it all to master again in about 2 months. Right now i can't support you building your own UI with our crates, since it's just not set up for use yet and really not done enough. However soon that will be possible.

**Comment by makepaddev** [comment]
We are in the last stretch to build the visual designtooling, and this might/will change our API's significantly. Once i can drag together an application with buttons and widgets i'll be more confident we got it right and at that point we'll document things.

**Comment by melMass** [comment]
Thanks!
I'll keep an eye out on when that will be possible!
Good luck on the last run!
